_posts/function-contracts.md

```diff
@@ -0,0 +1,409 @@
+
+What follows hereafter are a few more sections with additional information, such as the `modifies` clause used to reason about mutable memory, history expressions and a comparison with the stubbing feature we explored in an [earlier blog post](https://model-checking.github.io/kani-verifier-blog/2023/02/28/kani-internship-projects-2022-stubbing.html).
+
+---
+
+## A Neat Trick: Contracts for Non-local Functions
+
+Attributes for attaching function contracts only work on crate-local items but you may wish to stub an out of crate function. There is currently no builtin way to do so but you can achieve the same effect using a technique known as the "double stub". The idea is to first stub an external function to a local function that immediately calls the external function. Then the desired contract is attached to this local function which is then then stubbed as a `stub_verified` using the contract.
+
+```rust
+use external_crate::gcd;
+
+#[kani::ensures(result < max && result < min && result != 0)]
+fn local_gcd(max: i32, min: i32) -> i32 {
+    // immediate call with same arguments to the external function
+    gcd(max, min)
+}
+
+fn function_under_verification(...) {
+    ... gcd(...) ...
+}
+
+#[kani::proof]
+#[kani::stub(gcd, local_gcd)]
+#[kani::stub_verified(local_gcd)]
+fn harness() {
+    ... function_under_verification(...) ...
+}
+
+```
+
+The order of `stub` and `stub_verified` does not matter. With this technique we can attach a contract to the external `gcd` without having to change the code of `function_under_verification`. Unlike other uses of stubbing this does not pose a threat to soundness because the potentially unsound `stub` effectively replaces `gcd` with itself.
+
+## Soundness and Comparison with Stubbing
+
+Function contracts and stubbing are closely related as both allow the replacement of a costly computation with a cheap one via harness attributes. The crucial difference is that contracts preserve soundness. We will get into more detail later, but put simply: function contracts being sound means you can trust a successful verification.
+
+Conceptually stub `#[kani::stub(target, source)]` replaces all calls to `target` with `source` with weak constraints on what `source` is. `source` needs to be a function definition with a signature such that each call after the replacement typechecks. The details of what that means are unimportant, what matters is that no behavioral equivalence is checked beyond type compatibility.
+
+To illustrate this, lets consider an [example from XKCD](https://xkcd.com/221/) which stubs a `random` function by one that just returns `4`. This type of use of stubbing is completely legal and would not lead to a type error or verification failure.
+
+```rust
+// extern
+crate rand {
+    pub fn random<T>() -> T
+    where
+        Standard: Distribution<T>;
+}
```

Comment:
Mh, I get what you want to express here, but this is not really valid Ruts syntax. Maybe its better to just say `extern crate rand` with a comment that it exposes the `random` function?

Reply:
I've replaced with an empty/opaque body. I want the type signature in there to show that stubs don't even need to have the same trait bounds or types as what they replace so long as it still works with how the function actually ends up being called.